File: LeetCode150/C++/Merge Sorted Array.cpp
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        vector<int>res;
        if(n==0){
            return ;
        }

        int i=0;
        int j=0;
        while (j<m && i<n){
           
            if(nums1[j]>nums2[i]){
                res.push_back(nums2[i]);
                i++;
            }
            else{
                res.push_back(nums1[j]);
                j++;
            }
        }

        if(i==n){
            for(;j<m;j++){
                res.push_back(nums1[j]);
            }
        }

        if(j==m){
             for(;i<n;i++){
                res.push_back(nums2[i]);
            }
        }
        nums1=res;
    }
};


--------------------------------------------------------------------------------

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int i=m-1;
        int j=n-1;
        int k=n+m-1;
        
        while (j>=0 && i>=0){
           
            if(nums1[i]>nums2[j]){
                nums1[k]=nums1[i];
                i--;
            }
            else{
               nums1[k]=nums2[j];
                j--;
            }
            k--;
        }

        
            while (j>=0){
                nums1[k]=nums2[j];
                j--;
                k--;
            }
        
    }
};

File: LeetCode150/Python/Merge Sorted Array.py
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        res = []
        i = 0
        j = 0

        
        while i < m and j < n:
            if nums1[i] > nums2[j]:
                res.append(nums2[j])
                j += 1
            else:
                res.append(nums1[i])
                i += 1

        
        if i < m:
            res.extend(nums1[i:m])
        if j < n:
            res.extend(nums2[j:n])

        
        nums1[:] = res

//o(1) space start filling the biggest element from last 

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        i = m - 1
        j = n - 1
        k = m + n - 1

        while i >= 0 and j >= 0:
            if nums1[i] > nums2[j]:
                nums1[k] = nums1[i]
                i -= 1
            else:
                nums1[k] = nums2[j]
                j -= 1
            k -= 1

        
        while j >= 0:
            nums1[k] = nums2[j]
            j -= 1
            k -= 1


            



            
